firebase-perf/src/main/java/com/google/firebase/perf/util/Timer.java

```diff
@@ -14,92 +14,129 @@
 
 package com.google.firebase.perf.util;
 
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
 import android.os.Parcel;
 import android.os.Parcelable;
+import android.os.SystemClock;
 import androidx.annotation.NonNull;
 import com.google.android.gms.common.util.VisibleForTesting;
-import java.util.concurrent.TimeUnit;
 
-/** A Timer class provides both wall-clock (epoch) time and high resolution time (nano time). */
+/**
+ * A Timer class provides both wall-clock (epoch) time and monotonic time (elapsedRealtime).
+ * Timestamps are captured with millisecond-precision, because that's the time unit most widely
+ * available in Android APIs. However, private fields and public method returns are in microseconds
+ * due to Fireperf proto requirements.
+ */
 public class Timer implements Parcelable {
 
-  /** Wall-clock time or epoch time in microseconds, */
-  private long timeInMicros;
   /**
-   * High resolution time in nanoseconds. High resolution time should only be used to calculate
-   * duration or latency. It is not wall-clock time.
+   * Wall-clock time or epoch time in microseconds. Do NOT use for duration because wall-clock is
+   * not guaranteed to be monotonic: it can be set by the user or the phone network, thus it may
+   * jump forwards or backwards unpredictably. {@see SystemClock}
+   */
+  private long wallClockMicros;
+  /**
+   * Monotonic time measured in the {@link SystemClock#elapsedRealtime()} timebase. Only used to
+   * compute duration between 2 timestamps in the same timebase. It is NOT wall-clock time.
+   */
+  private long elapsedRealtimeMicros;
+
+  /**
+   * Returns a new Timer object as if it was stamped at the given elapsedRealtime. Uses current
+   * wall-clock as a reference to extrapolate the wall-clock at the given elapsedRealtime.
+   *
+   * @param elapsedRealtimeMillis timestamp in the {@link SystemClock#elapsedRealtime()} timebase
+   */
+  public static Timer ofElapsedRealtime(final long elapsedRealtimeMillis) {
+    long elapsedRealtimeMicros = MILLISECONDS.toMicros(elapsedRealtimeMillis);
+    long wallClockMicros = wallClockMicros() + (elapsedRealtimeMicros - elapsedRealtimeMicros());
+    return new Timer(wallClockMicros, elapsedRealtimeMicros);
+  }
+
+  /**
+   * Helper to get current wall-clock time from system API.
+   *
+   * @return wall-clock time in microseconds.
    */
-  private long highResTime;
+  private static long wallClockMicros() {
+    return MILLISECONDS.toMicros(System.currentTimeMillis());
+  }
 
   /**
-   * Construct Timer object using System clock. Make it package visible to be only accessible from
-   * com.google.firebase.perf.util.Clock.
+   * Helper to get current {@link SystemClock#elapsedRealtime()} from system API.
+   *
+   * @return wall-clock time in microseconds.
    */
+  private static long elapsedRealtimeMicros() {
+    return MILLISECONDS.toMicros(SystemClock.elapsedRealtime());
```

Comment:
We should use `elapsedRealtimeNanos` if the API level is >= 17

Reply:
Done